pkg/gui/controllers/branches_controller.go

```diff
@@ -178,75 +203,48 @@ func (self *BranchesController) setUpstream(selectedBranch *models.Branch) error
 					return self.c.Error(err)
 				}
 				return nil
-			},
-			Key: 'u',
+			})
 		},
-		{
-			LabelColumns: []string{self.c.Tr.SetUpstream},
-			OnPress: func() error {
-				return self.c.Helpers().Upstream.PromptForUpstreamWithoutInitialContent(selectedBranch, func(upstream string) error {
-					upstreamRemote, upstreamBranch, err := self.c.Helpers().Upstream.ParseUpstream(upstream)
-					if err != nil {
-						return self.c.Error(err)
-					}
+		Key: 's',
+	}
 
-					if err := self.c.Git().Branch.SetUpstream(upstreamRemote, upstreamBranch, selectedBranch.Name); err != nil {
-						return self.c.Error(err)
-					}
-					if err := self.c.Refresh(types.RefreshOptions{
-						Mode: types.SYNC,
-						Scope: []types.RefreshableView{
-							types.BRANCHES,
-							types.COMMITS,
-						},
-					}); err != nil {
-						return self.c.Error(err)
-					}
-					return nil
-				})
-			},
-			Key: 's',
+	upstream := lo.Ternary(selectedBranch.RemoteBranchStoredLocally(),
+		fmt.Sprintf("%s/%s", selectedBranch.UpstreamRemote, selectedBranch.Name),
+		self.c.Tr.UpstreamGenericName)
+	upstreamResetOptions := utils.ResolvePlaceholderString(
+		self.c.Tr.ViewUpstreamResetOptions,
+		map[string]string{"upstream": upstream},
+	)
+	upstreamResetTooltip := utils.ResolvePlaceholderString(
+		self.c.Tr.ViewUpstreamResetOptionsTooltip,
+		map[string]string{"upstream": upstream},
+	)
+
+	upstreamResetItem := &types.MenuItem{
+		LabelColumns: []string{upstreamResetOptions},
+		OpensMenu:    true,
+		OnPress: func() error {
+			err := self.c.Helpers().Refs.CreateGitResetMenu(upstream)
+			if err != nil {
+				return self.c.Error(err)
+			}
+			return nil
 		},
+		Tooltip: upstreamResetTooltip,
+		Key:     'g',
 	}
 
-	if selectedBranch.IsTrackingRemote() {
-		upstream := fmt.Sprintf("%s/%s", selectedBranch.UpstreamRemote, selectedBranch.Name)
-		upstreamResetOptions := utils.ResolvePlaceholderString(
-			self.c.Tr.ViewUpstreamResetOptions,
-			map[string]string{"upstream": upstream},
-		)
-		upstreamResetTooltip := utils.ResolvePlaceholderString(
-			self.c.Tr.ViewUpstreamResetOptionsTooltip,
-			map[string]string{"upstream": upstream},
-		)
-
-		options = append(options, &types.MenuItem{
-			LabelColumns: []string{upstreamResetOptions},
-			OpensMenu:    true,
-			OnPress: func() error {
-				if selectedBranch.RemoteBranchNotStoredLocally() {
-					return self.c.ErrorMsg(self.c.Tr.UpstreamNotStoredLocallyError)
-				}
+	if !selectedBranch.RemoteBranchStoredLocally() {
```

Comment:
It would be nice to introduce a builder pattern for these menu items where we can use `.DisableIf(remoteBranchNotStoredLocally, self.c.Tr.UpstreamNotSetError)` similar to what we do with constructing git args. Just thinking out loud, not requesting changes.

Reply:
I agree it would be nice, but I'm unsure how it would work. We could introduce it now, but we'd have to write
```
	(&types.MenuItem{
		Label: self.c.Tr.DeleteLocalBranch,
		Key:   'c',
		OnPress: func() error {
			return self.localDelete(branch)
		},
	}).DisableIf(checkedOutBranch.Name == branch.Name, self.c.Tr.CantDeleteCheckOutBranch)
```
with the parens, which looks awkward. Actually this is not much better than using `lo.Ternary` inside the braces.

I thought about introducing a `NewMenuItem` function, but since go doesn't have named arguments like Python does, calling it would be a pain; unless we also introduce a `NewMenuItemOpts` struct, but that would actually be identical to `MenuItem`, so that doesn't make sense either.

I feel like I'm overlooking something simple here...

Reply:
I was thinking something like:
```go
NewMenuItem().
  Label(self.c.Tr.DeleteLocalBranch).
  Key('c').
  OnPress(func() error {
	return self.localDelete(branch)
  )}.
  DisableIf(checkedOutBranch.Name == branch.Name, self.c.Tr.CantDeleteCheckOutBranch)
```

To be clear I think we should leave things as-is in this PR cos introducing this new pattern would take some thought